File: 3-leetcode/02-medium/36-valid-sudoku.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution
{
public:
  bool isValidSudoku(vector<vector<char>> &board)
  {
    int row[9][9] = {0};
    int col[9][9] = {0};
    int block[9][9] = {0};
    for (int r = 0; r < 9; r++)
    {
      for (int c = 0; c < 9; c++)
        if (board[r][c] != '.')
        {
          int curr = board[r][c] - '0' - 1;
          int b = r / 3 * 3 + c / 3;
          if (row[r][curr] || col[c][curr] || block[b][curr])
            return false;
          row[r][curr] = col[c][curr] = block[b][curr] = 1;
        }
    }

    return true;
  };
};

File: 3-leetcode/02-medium/js/36-valid-sudoku.js
const isValidSudoku = function (board) {
  let seen = new Set();
  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      let curr = board[r][c];
      let block = Math.floor(r / 3) * 3 + Math.floor(c / 3);
      if (board[r][c] !== '.') {
        if (
          seen.has(`${curr}r${r}`) ||
          seen.has(`${curr}c${c}`) ||
          seen.has(`${curr}b${block}`)
        )
          return false;
      }
      seen.add(`${curr}r${r}`);
      seen.add(`${curr}c${c}`);
      seen.add(`${curr}b${block}`);
    }
  }

  return true;
};
